integration/sandbox_run_rollback_test.go

```diff
@@ -229,6 +233,91 @@ func TestRunPodSandboxWithShimStartAndTeardownCNIFailure(t *testing.T) {
 	t.Run("JustCleanup", testCase(false))
 }
 
+// TestRunPodSandboxWithShimStartAndTeardownCNISlow should keep the sandbox
+// record if failed to rollback CNI API.
+func TestRunPodSandboxWithShimStartAndTeardownCNISlow(t *testing.T) {
+	if runtime.GOOS != "linux" {
+		t.Skip()
+	}
+	if os.Getenv("ENABLE_CRI_SANDBOXES") != "" {
+		t.Skip()
+	}
+
+	t.Log("Init PodSandboxConfig with specific key")
+	labels := map[string]string{
+		t.Name(): "true",
+	}
+	sbConfig := PodSandboxConfig(t.Name(), "failpoint", WithPodLabels(labels))
+
+	t.Log("Inject CNI failpoint")
+	conf := &failpointConf{
+		// Delay 10s
+		Add: "1*delay(10000)",
+		// This is to make sure the sandbox container won't be deleted even if the defer functions are invoked in sandbox_run.go
+		// Without this, the test is flaky because ListPodSandbox may return empty list.
+		Del: "1*error(please retry)",
+	}
+	injectCNIFailpoint(t, sbConfig, conf)
+
+	go func() {
+		t.Log("Create a sandbox")
+		_, err := runtimeService.RunPodSandbox(sbConfig, failpointRuntimeHandler)
+		require.Error(t, err)
+		require.ErrorContains(t, err, "error reading from server: EOF")
+	}()
+
+	time.Sleep(5 * time.Second)
+	t.Log("Restart containerd after 5s")
+	RestartContainerd(t)
+
+	t.Log("ListPodSandbox with the specific label")
+	l, err := runtimeService.ListPodSandbox(&criapiv1.PodSandboxFilter{LabelSelector: labels})
+	require.NoError(t, err)
+	require.Len(t, l, 1)
+
+	sb := l[0]
+
+	defer func() {
+		t.Log("Cleanup leaky sandbox")
+		runtimeService.StopPodSandbox(sb.Id)
+		runtimeService.RemovePodSandbox(sb.Id)
+	}()
+
+	require.Equal(t, sb.State, criapiv1.PodSandboxState_SANDBOX_NOTREADY)
+	require.Equal(t, sb.Metadata.Name, sbConfig.Metadata.Name)
+	require.Equal(t, sb.Metadata.Namespace, sbConfig.Metadata.Namespace)
+	require.Equal(t, sb.Metadata.Uid, sbConfig.Metadata.Uid)
+	require.Equal(t, sb.Metadata.Attempt, sbConfig.Metadata.Attempt)
```

Comment:
I'd generally recommend `require` be used for situations where continuing the test would cause a panic and `assert` be used for other errors.  The value of `assert` is that the test can continue and you can see all failed assertions, rather than `require` which fails the test immediately (this is the difference between `t.Error` and `t.Fatal`).

```suggestion
	assert.Equal(t, sb.State, criapiv1.PodSandboxState_SANDBOX_NOTREADY)
	assert.Equal(t, sb.Metadata.Name, sbConfig.Metadata.Name)
	assert.Equal(t, sb.Metadata.Namespace, sbConfig.Metadata.Namespace)
	asssert.Equal(t, sb.Metadata.Uid, sbConfig.Metadata.Uid)
	assert.Equal(t, sb.Metadata.Attempt, sbConfig.Metadata.Attempt)
```

Reply:
Good to know. I agree with this good practice. Let me update.